relentless/simulate/generic.py

```diff
@@ -0,0 +1,301 @@
+import importlib
+import inspect
+
+import numpy as np
+
+from . import simulate
+
+class SimulationOperationAdapter(simulate.SimulationOperation):
+    """
+    """
+    backends = {}
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+        self._op = None
+        self._backend = None
+
+    def __call__(self, sim):
+        if self._op is None or self._backend != sim.backend:
+            backend = self.backends.get(sim.backend)
+            if not backend:
+                raise TypeError('Simulation backend {} not registered.'.format(backend))
+
+            op_name = type(self).__name__
+            try:
+                BackendOp = getattr(backend,op_name)
+            except AttributeError:
+                raise TypeError('{}.{}.{} operation not found.'.format(backend.__module__,backend.__name__,op_name))
+
+            self._op = BackendOp(*self.args,**self.kwargs)
+            self._backend = sim.backend
+
+        return self._op(sim)
+
```

Comment:
Making a note to myself: we should find a way to set (and forward) attributes... this is tricky because the backend `_op` is not constructed until `__call__`. If `__call__` has been executed once, no problem, but the more common case is that it has not. Probably we need to track what properties get set on the adapter class, then loop over them to forward them to the backend op in call when it gets created.